### DISCUSSION TITLE
리플랙션 declared 메서드

### DISCUSSION THREAD

**Discussion by 0sunset0** [cowaine/is-my-code-that-strange#75] [Chapter 09-10 (악마, 이름 설계)] [answered]
```java
Field field1 = Level.class.getDeclaredField("value");
Field field2 = Level.class.getField("value");
```

리플랙션 api에서 getField()와 getDeclaredField()처럼 declared가 없는 메서드와 있는 메서드의 차이는 무엇일까요?

**Comment by Irisation23** [comment, marked as answer]
<img width="934" alt="image" src="https://github.com/user-attachments/assets/ea35fe50-f6ce-4361-aa70-38d8b06eac49">


### getField()

- 해당 클래스의 public 필드 (**부모 클래스 포함**)을 얻어온다.

### getDeclaredField()

- 해당 클래스의 접근제어자를 무시한 모든 필드를 얻어온다. (단 해당 클래스에 선언된 것만)
- 단 부모 클래스의 접근제어자를 무시한 모든 필드를 얻어오는 방법은 해당 클래스를 재귀적으로 탐색을 수행하여 얻어올 수 있음.

``` java
private static final String EMPTY = null;

public static Field getFieldUp(Class clazz, String name) {
  if (clazz == EMPTY) {
    return EMPTY; 
  }
  Field[] fields = clazz.getDeclaredFields();
  for (Field field : fields) {
    if (field.getName().equals(name)) {
      return field;
    }
  }
  
Class parentClass = clazz.getSuperclass();

  if (parentClass == EMPTY) {
    return EMPTY;
  }

  return getFieldUp(parentClass, name); // 해당 메서드 재귀 호출
}
```

> 참조 
> [Java class getField vs getDeclaredField](https://docs.oracle.com/javase/tutorial/reflect/class/classMembers.html)
> [reflect-class-classMembers](https://medium.com/@narusas/java-class-getfield-vs-getdeclaredfield-1a2fdb2d0b27)
